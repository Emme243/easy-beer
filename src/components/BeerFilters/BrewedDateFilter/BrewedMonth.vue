<template>
  <v-dialog
    v-model="isModalOpen"
    overlay-opacity="0.3"
    :close-on-content-click="false"
    transition="scale-transition"
    offset-y
    max-width="290px"
    min-width="auto"
  >
    <template v-slot:activator="{ on, attrs }">
      <v-text-field
        :value="brewedMonth"
        :label="inputLabel"
        readonly
        :append-icon="brewedMonth && 'fa fa-arrow-rotate-left'"
        v-bind="attrs"
        v-on="on"
        @click:append="handleBrewedMonthChange('')"
      ></v-text-field>
    </template>
    <v-date-picker
      type="month"
      scrollable
      @input="handleBrewedMonthChange"
      :value="brewedMonth"
      :min="brewedMinMonth"
      :max="brewedMaxMonth"
      show-current="2014-01"
    >
      <v-spacer></v-spacer>
      <v-btn text color="primary" @click="handleBrewedMonthChange('')"> Limpiar </v-btn>
      <v-btn text color="primary" @click="isModalOpen = false"> OK </v-btn>
    </v-date-picker>
  </v-dialog>
</template>

<script>
import { mapActions } from 'vuex';

export default {
  name: 'BrewedMonth',
  props: {
    inputLabel: {
      type: String,
      required: true,
    },
    brewedMonth: {
      type: String,
      required: true,
    },
    brewedMinMonth: {
      type: String,
      default: '',
    },
    brewedMaxMonth: {
      type: String,
      default: '',
    },
  },
  data: () => ({ isModalOpen: false }),
  methods: {
    ...mapActions({ resetPageInStore: 'filter/resetPage' }),
    handleBrewedMonthChange(month) {
      this.$emit('setBrewedMonthInStore', month);
      this.resetPageInStore();
    },
  },
};
</script>

<style scoped></style>
