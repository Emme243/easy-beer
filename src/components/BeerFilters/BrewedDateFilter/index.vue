<template>
  <div>
    <div class="d-flex align-center">
      <h3 class="text-h6 mr-4">Fecha de elaboración</h3>
      <!--Beer Brewed Date Reset Button-->
      <button
        @click="handleResetBrewedDate"
        v-show="!!initialBrewedMonth || !!finalBrewedMonth"
        data-test-id="reset-brewed-date-btn"
      >
        <v-icon small>fa fa-arrow-rotate-left</v-icon>
      </button>
    </div>
    <v-row>
      <v-col cols="6">
        <InitialBrewedMonth />
      </v-col>
      <v-col cols="6">
        <FinalBrewedMonth />
      </v-col>
    </v-row>
    <span class="font-weight-bold" data-test-id="info-message">
      Se mostrarán cervezas elaboradas {{ helperText }}
    </span>
  </div>
</template>

<script>
import FinalBrewedMonth from '@/components/BeerFilters/BrewedDateFilter/FinalBrewedMonth';
import InitialBrewedMonth from '@/components/BeerFilters/BrewedDateFilter/InitialBrewedMonth';
import { mapGetters } from 'vuex';
import { mapActions } from 'vuex';

export default {
  name: 'BrewedDateFilter',
  components: { FinalBrewedMonth, InitialBrewedMonth },
  methods: {
    ...mapActions({
      resetBrewedDateInStore: 'filter/resetBrewedDate',
      resetPageInStore: 'filter/resetPage',
    }),
    handleResetBrewedDate() {
      this.resetBrewedDateInStore();
      this.resetPageInStore();
    },
  },
  computed: {
    ...mapGetters({
      initialBrewedMonth: 'filter/initialBrewedMonth',
      finalBrewedMonth: 'filter/finalBrewedMonth',
    }),
    helperText() {
      if (!this.initialBrewedMonth && !this.finalBrewedMonth) return 'hasta la fecha actual';
      if (this.initialBrewedMonth === this.finalBrewedMonth) return `en ${this.initialBrewedMonth}`;
      if (this.initialBrewedMonth && !this.finalBrewedMonth)
        return `después de ${this.initialBrewedMonth}`;
      if (!this.initialBrewedMonth && this.finalBrewedMonth)
        return `antes de ${this.finalBrewedMonth}`;
      if (this.initialBrewedMonth && this.initialBrewedMonth)
        return `entre ${this.initialBrewedMonth} y ${this.finalBrewedMonth}`;
    },
  },
};
</script>

<style scoped></style>
