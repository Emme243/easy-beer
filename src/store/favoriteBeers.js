export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
