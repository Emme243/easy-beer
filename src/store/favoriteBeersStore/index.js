export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
