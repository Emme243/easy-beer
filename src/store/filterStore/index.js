import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
