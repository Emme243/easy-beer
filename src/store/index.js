import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
