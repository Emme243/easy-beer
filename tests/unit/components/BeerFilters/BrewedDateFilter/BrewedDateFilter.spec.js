import BrewedDateFilter from '@/components/BeerFilters/BrewedDateFilter';
import store from '@/store';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';

describe('Pruebas en el componente BrewedDateFilter/index.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const mountFunction = options => {
    return shallowMount(BrewedDateFilter, {
      localVue,
      vuetify,
      store,
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const resetBrewedDateBtnSelector = '[data-test-id="reset-brewed-date-btn"]';
  const infoMessageSelector = '[data-test-id="info-message"]';

  it('Debe hacer match con el snapshot', function () {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('El botón de reseteo de fechas debe emitir el dispatch resetBrewedDate y el dispatch resetPage cuando se hace click', async function () {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const resetBrewedDateBtn = wrapper.find(resetBrewedDateBtnSelector);
    expect(resetBrewedDateBtn.exists()).toBeTruthy();
    await resetBrewedDateBtn.trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetBrewedDate');
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetPage');
    expect(store.dispatch).toHaveBeenCalledTimes(2);
  });

  it('El mensaje de información debe decir `Se mostrarán cervezas elaboradas hasta la fecha actual` cuando initialBrewedMonth y finalBrewedMonth no tengan un valor', function () {
    const wrapper = mountFunction({
      computed: {
        initialBrewedMonth: () => '',
        finalBrewedMonth: () => '',
      },
    });
    const infoMessage = wrapper.find(infoMessageSelector);
    expect(infoMessage.exists()).toBeTruthy();
    expect(infoMessage.text()).toBe('Se mostrarán cervezas elaboradas hasta la fecha actual');
  });

  it('El mensaje de información debe decir `Se mostrarán cervezas elaboradas entre ${initialBrewedMonth} y ${finalBrewedMonth}` cuando initialBrewedMonth y finalBrewedMonth tengan un valor y sean diferentes', function () {
    const wrapper = mountFunction({
      computed: {
        initialBrewedMonth: () => '02-2014',
        finalBrewedMonth: () => '02-2015',
      },
    });
    const infoMessage = wrapper.find(infoMessageSelector);
    expect(infoMessage.exists()).toBeTruthy();
    expect(infoMessage.text()).toBe('Se mostrarán cervezas elaboradas entre 02-2014 y 02-2015');
  });

  it('El mensaje de información debe decir `Se mostrarán cervezas elaboradas en ${initialBrewedMonth}` cuando initialBrewedMonth y finalBrewedMonth tengan un valor y sean iguales', function () {
    const wrapper = mountFunction({
      computed: {
        initialBrewedMonth: () => '02-2014',
        finalBrewedMonth: () => '02-2014',
      },
    });
    const infoMessage = wrapper.find(infoMessageSelector);
    expect(infoMessage.exists()).toBeTruthy();
    expect(infoMessage.text()).toBe('Se mostrarán cervezas elaboradas en 02-2014');
  });

  it('El mensaje de información debe decir `Se mostrarán cervezas elaboradas después de ${initialBrewedMonth}` cuando initialBrewedMonth tenga un valor y finalBrewedMonth no tenga un valor', function () {
    const wrapper = mountFunction({
      computed: {
        initialBrewedMonth: () => '02-2014',
        finalBrewedMonth: () => '',
      },
    });
    const infoMessage = wrapper.find(infoMessageSelector);
    expect(infoMessage.exists()).toBeTruthy();
    expect(infoMessage.text()).toBe('Se mostrarán cervezas elaboradas después de 02-2014');
  });

  it('El mensaje de información debe decir `Se mostrarán cervezas elaboradas antes de ${finalBrewedMonth}` cuando initialBrewedMonth no tenga un valor y finalBrewedMonth tenga un valor', function () {
    const wrapper = mountFunction({
      computed: {
        initialBrewedMonth: () => '',
        finalBrewedMonth: () => '02-2014',
      },
    });
    const infoMessage = wrapper.find(infoMessageSelector);
    expect(infoMessage.exists()).toBeTruthy();
    expect(infoMessage.text()).toBe('Se mostrarán cervezas elaboradas antes de 02-2014');
  });
});
