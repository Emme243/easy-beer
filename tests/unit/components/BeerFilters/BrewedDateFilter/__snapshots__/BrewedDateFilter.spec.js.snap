// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente BrewedDateFilter/index.vue Debe hacer match con el snapshot 1`] = `
<div>
  <div class="d-flex align-center">
    <h3 class="text-h6 mr-4">Fecha de elaboración</h3> <button data-test-id="reset-brewed-date-btn" style="display: none;">
      <v-icon-stub small="">fa fa-arrow-rotate-left</v-icon-stub>
    </button>
  </div>
  <v-row-stub tag="div">
    <v-col-stub cols="6" tag="div">
      <initialbrewedmonth-stub></initialbrewedmonth-stub>
    </v-col-stub>
    <v-col-stub cols="6" tag="div">
      <finalbrewedmonth-stub></finalbrewedmonth-stub>
    </v-col-stub>
  </v-row-stub> <span data-test-id="info-message" class="font-weight-bold">
    Se mostrarán cervezas elaboradas hasta la fecha actual
  </span>
</div>
`;
